src/code/z_map_exp.c

```diff
@@ -1,31 +1,259 @@
+
+    func_80105C40(room);
+
+    if (room >= 0) {
+        if (Map_IsInDungeonOrBossArea(play)) {
+            gSaveContext.save.permanentSceneFlags[Play_GetOriginalSceneNumber(play->sceneNum)].rooms |= gBitFlags[room];
+            interfaceCtx->mapRoomNum = room;
+            interfaceCtx->unk_27A = mapIndex;
+        }
+    } else {
+        interfaceCtx->mapRoomNum = 0;
+    }
+
+    if (gSaveContext.sunsSongState != SUNSSONG_SPEED_TIME) {
+        gSaveContext.sunsSongState = SUNSSONG_INACTIVE;
+    }
+}
+
+void Map_Destroy(void) {
+    func_80105A40();
+}
+
+void Map_Init(PlayState* play) {
+    InterfaceContext* interfaceCtx = &play->interfaceCtx;
+    s32 dungeonIndex;
+
+    func_80105C40(play->roomCtx.currRoom.num);
+    interfaceCtx->unk_278 = -1;
+    interfaceCtx->unk_27A = -1;
+    interfaceCtx->mapSegment = THA_AllocEndAlign16(&play->state.heap, 0x1000);
+    if (func_8010A2AC(play)) {
+        gSaveContext.mapIndex = func_8010A238(play);
+        return;
+    }
+
+    if (Map_IsInDungeonOrBossArea(play)) {
+        dungeonIndex = Map_GetDungeonOrBossAreaIndex(play);
+        gSaveContext.mapIndex = dungeonIndex;
+        switch (play->sceneNum) {
+            case SCENE_MITURIN_BS:
+                dungeonIndex = DUNGEON_INDEX_WOODFALL_TEMPLE;
+                break;
+
+            case SCENE_HAKUGIN_BS:
+                dungeonIndex = DUNGEON_INDEX_SNOWHEAD_TEMPLE;
+                break;
+
+            case SCENE_SEA_BS:
+                dungeonIndex = DUNGEON_INDEX_GREAT_BAY_TEMPLE;
+                break;
+
+            case SCENE_INISIE_BS:
+                dungeonIndex = DUNGEON_INDEX_STONE_TOWER_TEMPLE;
+                break;
+        }
+
+        gSaveContext.dungeonIndex = dungeonIndex;
+        Map_InitRoomData(play, play->roomCtx.currRoom.num);
+    }
+}
+
+void Minimap_Draw(PlayState* play) {
+    func_80106644(play, sMinimapInitPosX, sMinimapInitPosZ, sMinimapInitDir);
+}
+
+s16 sLastRoomNum = 99;
+
+#define FLOOR_INDEX_MAX 4
+
+void Map_Update(PlayState* play) {
+    InterfaceContext* interfaceCtx = &play->interfaceCtx;
+    Player* player = GET_PLAYER(play);
+    Input* controller = CONTROLLER1(&play->state);
+    s32 pad1;
+    s16 floor;
+    s32 pad2;
+
+    if ((play->pauseCtx.state < 4) && (CHECK_BTN_ALL(controller->press.button, BTN_L)) && (!Play_InCsMode(play)) &&
+        (!func_80106530(play))) {
+        if (!R_MINIMAP_DISABLED) {
+            play_sound(NA_SE_SY_CAMERA_ZOOM_UP);
+        } else {
+            play_sound(NA_SE_SY_CAMERA_ZOOM_DOWN);
+        }
+
+        R_MINIMAP_DISABLED ^= 1;
+    }
+
+    func_80105B34(play);
+
+    if ((play->pauseCtx.state == 0) && (play->pauseCtx.debugState == 0)) {
+        if (Map_IsInDungeonArea(play)) {
+            floor = func_80109124(player->actor.world.pos.y);
+            if (floor != -1) {
+                gSaveContext.save.permanentSceneFlags[Play_GetOriginalSceneNumber(play->sceneNum)].unk_14 |=
+                    gBitFlags[FLOOR_INDEX_MAX - floor];
```

Comment:
```suggestion
                gSaveContext.save.permanentSceneFlags[Play_GetOriginalSceneNumber(play->sceneNum)].floors |=
                    gBitFlags[FLOOR_INDEX_MAX - floor];
```
Same here, this was missed.

Reply:
This gets used in quite a few actors for non-floor-like things, seemingly. I've made the change so you can see for yourself, but I'm curious if you're still up for it once you see it.

Reply:
Interesting, so they do seem to reuse `unk_14` based on scene. For all dungeon scenes, it's used to store floors bits. But for non-dungeons scenes, it's used for various other bits depending on the scene. So yeah, it'll either need a more generic name or a union of flags. Definitely outside the scope of this PR, so yeah can keep it as `unk_14` for now.

Reply:
Sounds good, I'll change it back.